### PR TITLE
Player death

### PR DESCRIPTION
**Pull request by samuelWilliams99**
Adds a player death box containing all a players items when they die, also clearing their inventory.
Also add a respawn delay.

Blocked until party system is merged

- [x] Model for death box

### CODE REVIEW

File: entities/entities/rvr_death_box/init.lua [hunk 1]
```diff
@@ -0,0 +1,76 @@
+AddCSLuaFile( "cl_init.lua" )
+AddCSLuaFile( "shared.lua" )
+
+include( "shared.lua" )
+
+hook.Add( "InitPostEntity", "RVR_FindWaterLevel", function()
+    local trace = util.TraceLine( {
+        start = Vector( 0, 0, 10000 ),
```
Comment (overlordbleck) [inline entities/entities/rvr_death_box/init.lua]: Can we just use `math.huge` for these things?

File: entities/entities/rvr_death_box/init.lua [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+function ENT:Initialize()
+    self.BaseClass.Initialize( self )
+
+    -- Set up despawn timer
+    local despawnTime = GAMEMODE.Config.PlayerDeath.DEATH_BOX_DESPAWN_TIME
+    if despawnTime > 0 then
+        self.timerIdentifier = "rvr_death_box_despawn_" .. self:EntIndex()
+        local this = self
+
+        timer.Create( self.timerIdentifier, despawnTime, 1, function()
```
Comment (overlordbleck) [inline entities/entities/rvr_death_box/init.lua]:
```suggestion
        timer.Create( this.timerIdentifier, despawnTime, 1, function()
```

File: gamemode/modules/inventory/sv_inventory.lua [hunk 3]
```diff
@@ -179,53 +210,65 @@ function inv.slotCanContain( ent, position, item )
...
-function inv.attemptPickupItem( ply, item, count )
-    if not ply.RVR_Inventory then return false, 0 end
+function inv.attemptPickupItem( ent, item, count )
+    if not ent.RVR_Inventory then return false, 0 end
     count = count or 1
     local originalCount = count
 
+    if ent.RVR_Inventory.PreventAdding then return false, 0 end
+
+    local isPlayer = type( ent ) == "Player"
```
Comment (overlordbleck) [inline gamemode/modules/inventory/sv_inventory.lua]:
```suggestion
    local isPlayer = ent:IsPlayer()
```

File: gamemode/modules/player_death/cl_playerdeath.lua [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        surface.DrawTexturedRect( iconX - radius * checkmarkSize, iconY - radius * checkmarkSize,
+            radius * checkmarkSize * 2, radius * checkmarkSize * 2 )
+    end
+end )
+
+hook.Add( "HUDPaintBackground", "RVR_PlayerDeathScreen", function()
+    if LocalPlayer():Alive() then return end
+    surface.SetDrawColor( 255, 0, 0, 70 )
+    surface.DrawRect( 0, 0, ScrW(), ScrH() )
+end )
```
Comment (overlordbleck) [inline gamemode/modules/player_death/cl_playerdeath.lua]:
```suggestion
end )

```

**Review by overlordbleck — COMMENTED** [review]
WOW SAM FUCKING GREAT JOB.

File: entities/entities/rvr_death_box/init.lua (hunk 1 above)
Comment (samuelWilliams99) [reply entities/entities/rvr_death_box/init.lua]: Can for the second one, not the first. Also this is kind of a shit solution to finding the water surface height, but I can't think of anything better

Comment (samuelWilliams99) [reply entities/entities/rvr_death_box/init.lua]: Wait no I can't, it returns 0/0 which isnt exactly what i was going for

File: gamemode/modules/hud/cl_hud.lua [hunk 5]
```diff
@@ -32,44 +32,46 @@ local stats = {
 function GM:HUDPaint()
     hook.Run( "HUDDrawTargetID" )
 
-    local barHeight = ScrH() * 0.04
+    if hook.Run( "HUDShouldDraw", "RVR_Stats" ) then
```
Comment (plally) [inline gamemode/modules/hud/cl_hud.lua]: you could put this in its own function `drawStats`

File: gamemode/modules/inventory/sv_inventory.lua [hunk 6]
```diff
@@ -24,6 +24,37 @@ end
...
+    end
+end
+
+function inv.isEmpty( ent )
+    local entInv = ent.RVR_Inventory
+    if not entInv then return true end
+
+    if not table.IsEmpty( entInv.Inventory ) then return false end
+
+    if type( ent ) == "Player" then
```
Comment (plally) [inline gamemode/modules/inventory/sv_inventory.lua]: if `ent:IsPlayer`?

File: entities/entities/rvr_death_box/init.lua (hunk 1 above)
Comment (plally) [reply entities/entities/rvr_death_box/init.lua]: would this break if the map isn't tall enough?

File: gamemode/modules/inventory/sv_inventory.lua [hunk 7]
```diff
@@ -24,6 +24,37 @@ end
 
 hook.Add( "PlayerInitialSpawn", "RVR_SetupInventory", inv.setupPlayer )
 
+function inv.clearInventory( ent )
+    if not ent.RVR_Inventory then return end
+
+    ent.RVR_Inventory.Inventory = {}
+
+    if type( ent ) == "Player" then
```
Comment (plally) [inline gamemode/modules/inventory/sv_inventory.lua]: `ent:IsPlayer()`?

File: entities/entities/rvr_death_box/init.lua (hunk 1 above)
Comment (overlordbleck) [reply entities/entities/rvr_death_box/init.lua]: I thought that's why the mask existed

Comment (samuelWilliams99) [reply entities/entities/rvr_death_box/init.lua]: I think it will, need something better

**Pull request merged**